pkg/issuer/venafi/client/request.go

```diff
@@ -102,60 +116,73 @@ func (v *Venafi) Sign(csrPEM []byte, duration time.Duration, customFields []inte
+	if err != nil {
+		return nil, err
+	}
+	vreq.FriendlyName = friendlyName
+
 	// Set options on the request
 	vreq.CsrOrigin = certificate.UserProvidedCSR
-	//// TODO: better set the timeout here. Right now, we'll block for this amount of time.
-	vreq.Timeout = time.Minute * 5
-
-	// Set the 'ObjectName' through the request friendly name. This is set in
-	// order of precedence CN->DNS->URI.
-	switch {
-	case len(tmpl.Subject.CommonName) > 0:
-		vreq.FriendlyName = tmpl.Subject.CommonName
-		break
-	case len(tmpl.DNSNames) > 0:
-		vreq.FriendlyName = tmpl.DNSNames[0]
-		break
-	case len(tmpl.URIs) > 0:
-		vreq.FriendlyName = tmpl.URIs[0].String()
-		break
-	default:
-		return nil, errors.New(
-			"certificate request contains no Common Name, DNS Name, nor URI SAN, at least one must be supplied to be used as the Venafi certificate objects name")
-	}
 
 	// Set the request CSR with the passed value
 	if err := vreq.SetCSR(csrPEM); err != nil {
 		return nil, err
 	}
 
-	// Send the certificate signing request to Venafi
-	requestID, err := v.client.RequestCertificate(vreq)
-	if err != nil {
-		return nil, err
-	}
+	return vreq, nil
+}
 
-	// Set the PickupID so vcert does not have to look it up by the fingerprint
-	vreq.PickupID = requestID
+func convertCustomFieldsToVcert(customFields []api.CustomField) ([]certificate.CustomField, error) {
+	var out []certificate.CustomField
+	if len(customFields) > 0 {
+		for _, field := range customFields {
+			var fieldType certificate.CustomFieldType
+			switch field.Type {
+			case api.CustomFieldTypePlain, "":
+				fieldType = certificate.CustomFieldPlain
+				break
+			default:
+				return nil, ErrCustomFieldsType{Type: field.Type}
+			}
 
-	// Retrieve the certificate from request
-	pemCollection, err := v.client.RetrieveCertificate(vreq)
-	if err != nil {
-		return nil, err
+			out = append(out, certificate.CustomField{
+				Type:  fieldType,
+				Name:  field.Name,
+				Value: field.Value,
+			})
+		}
 	}
 
-	// Construct the certificate chain and return the new keypair
-	cs := append([]string{pemCollection.Certificate}, pemCollection.Chain...)
-	chain := strings.Join(cs, "\n")
-
-	return []byte(chain), nil
+	return out, nil
 }
 
 func newVRequest(cert *x509.Certificate) *certificate.Request {
 	req := certificate.NewRequest(cert)
 
 	// overwrite entire Subject block
 	req.Subject = cert.Subject
+	// Add cert-manager origin tag
+	req.CustomFields = []certificate.CustomField{
+		{
+			Type:  certificate.CustomFieldOrigin,
+			Value: "Jetstack cert-manager",
+		},
+	}
 	return req
 }
+
+func getVcertFriendlyName(crt *x509.Certificate) (string, error) {
+	// Set the 'ObjectName' through the vcert friendly name. This is set in
+	// order of precedence CN->DNS->URI.
+	switch {
+	case len(crt.Subject.CommonName) > 0:
+		return crt.Subject.CommonName, nil
+	case len(crt.DNSNames) > 0:
+		return crt.DNSNames[0], nil
+	case len(crt.URIs) > 0:
+		return crt.URIs[0].String(), nil
```

Comment:
Do we need to account for EmailSANs too? Any other SAN types?

Reply:
Added email and IP, these were not present when this was originally written I think...